app/models/sftp/cli.rb

```diff
@@ -0,0 +1,294 @@
+      error_msg = output if error_msg.empty?
+      raise StatusException, "SFTP command failed: #{command}\n#{error_msg}"
+    end
+
+    # Builds the command parts as an array for safe execution with Open3.
+    # Uses a temporary batch file to pass commands to sftp, which is safer than heredoc.
+    # The batch file is tracked and cleaned up in the cleanup method.
+    def build_sftp_command_parts(sftp_commands)
+      commands = Array(sftp_commands).join("\n")
+      batch_file = Tempfile.new('sftp_batch')
+      batch_file.write("#{commands}\nquit\n")
+      batch_file.close
+      File.chmod(0o600, batch_file.path)
+      @batch_files << batch_file
+
+      build_base_command_parts(include_batch_file: batch_file.path)
+    end
+
+    # Builds the base command parts as an array for safe execution.
+    def build_base_command_parts(include_batch_file: nil)
+      ensure_password_file if password_auth?
+
+      sftp_parts = ['sftp']
+      sftp_parts << '-P' << @options[:port].to_s if @options[:port]
+      if key_auth?
+        sftp_parts << '-i' << @options[:keys].first
+        sftp_parts << '-o' << 'PreferredAuthentications=publickey' if @options[:keys_only]
+      end
+      sftp_parts.concat(build_ssh_options_parts)
+      # -b flag must come before the destination
+      sftp_parts << '-b' << include_batch_file if include_batch_file
+      sftp_parts << %(#{@username}@#{@host})
+
+      if password_auth?
+        ['sshpass', '-f', @password_file.path] + sftp_parts
+      else
+        sftp_parts
+      end
+    end
+
+    # Builds SSH option flags as separate -o flag=value pairs for the sftp command.
+    def build_ssh_options_parts
+      opts = []
+      opts << '-o' << 'PasswordAuthentication=yes' if password_auth?
+      opts << '-o' << 'StrictHostKeyChecking=no' if @options[:append_all_supported_algorithms]
+      opts
+    end
+
+    # Returns true if password authentication should be used.
+    def password_auth?
+      @options[:password].present? || @options[:auth_methods]&.include?('password')
+    end
+
+    # Returns true if SSH key authentication should be used.
+    def key_auth?
+      @options[:keys].present? && @options[:keys].any?
+    end
+
+    # Creates a temporary file with restricted permissions (0600) containing the password.
+    # This file is used by sshpass to avoid exposing the password in process lists or command history.
+    def ensure_password_file
+      return if @password_file && File.exist?(@password_file.path)
+
+      @password_file = Tempfile.new('sftp_password')
+      @password_file.write(@options[:password] || '')
+      @password_file.close
+      File.chmod(0o600, @password_file.path)
+    end
+
+    # Escapes special characters in file paths for safe use in shell commands.
+    def escape_path(path)
+      path.to_s.gsub(' ', '\\ ').gsub('(', '\\(').gsub(')', '\\)').gsub('[', '\\[').gsub(']', '\\]').gsub('*', '\\*').gsub('?', '\\?')
+    end
```

Comment:
I think this should be
```suggestion
   def escape_path(path)
    Shellwords.escape(path.to_s)
  end
```